packages/lesswrong/server/authentication/auth0Accounts.ts

```diff
@@ -0,0 +1,41 @@
+import { Profile } from "passport-auth0";
+import Users from "../../lib/vulcan-users";
+import { slugify, Utils } from "../../lib/vulcan-lib/utils";
+import { updateMutator } from "../vulcan-lib/mutators";
+
+export async function mergeAccountWithAuth0(user: DbUser, profile: Profile) {
+  return await updateMutator({
+    collection: Users,
+    documentId: user._id,
+    // This is the correct way to set a nested property according to Mongo, but
+    // it's very hard to get it to type correctly. TS thinks we're setting a
+    // completely different field, `services.auth0`, not setting a nested one.
+    set: {'services.auth0': profile} as any,
+    // Normal updates are not supposed to update services
+    validate: false
+    // TODO: Soon we should delete passwords, and maybe(?) resume tokens when we do this
+  })
+}
+
+export async function userFromAuth0Profile(profile: Profile): Promise<Partial<DbUser>> {
+  const email = profile.emails?.[0].value
+  const displayNameMatchesEmail = email === profile.displayName
+  const displayName = displayNameMatchesEmail ?
+    `new_user_${Math.floor(Math.random() * 10e9)}` :
+    profile.displayName
+  return {
+    email,
+    emails: email ?
+      [{
+        address: email,
+        verified: !!profile._json.email_verified
```

Comment:
If I'm understanding correctly, this potentially creates a problem if the user creates a new Auth0 account, creates an EA Forum account linked to it, then verifies their Auth0 account's email address, in that order. Then the EA Forum account won't have a verified email address? It might be that they get notifications telling them to verify their email address (through EA Forum) and it's fine. But I'd be sure to test this case specifically.

Reply:
This whole flow is a TODO for the moment.